File: python/turingTransitionFunction.py
from turingTape import TuringTapeMove
from turingTyping import TuringState, TuringSymbol, raiseIfInvalidState, raiseIfInvalidSymbol

class TuringTransitionFunction():
    _haltState:TuringState
    _transitions:dict[tuple[TuringState, TuringSymbol], tuple[TuringState,TuringSymbol,TuringTapeMove]]

    def __init__(self, haltState:TuringState) -> None:
        raiseIfInvalidState(haltState)
        self._haltState = haltState
        self._transitions = {}
    
    def set(self, frm:tuple[TuringState, TuringSymbol], to:tuple[TuringState,TuringSymbol,TuringTapeMove]) -> None:
        raiseIfInvalidState(frm[0])
        raiseIfInvalidSymbol(frm[1])
        raiseIfInvalidState(to[0])
        raiseIfInvalidSymbol(to[1])

        if frm in self._transitions.keys():
            raise ValueError("From setup has already been defined")
        
        self._transitions[frm] = to
    
    def get(self, state:TuringState, symbol:TuringSymbol) -> tuple[TuringState,TuringSymbol,TuringTapeMove]:
        raiseIfInvalidState(state)
        raiseIfInvalidSymbol(symbol)
    
        frm:tuple[TuringState,TuringSymbol] = (state, symbol)
        if frm not in self._transitions.keys():
            return (self._haltState, symbol, TuringTapeMove.Stay)
        return self._transitions[frm]
        

File: cpp/turing-transition-function.cpp
#include <map>
#include <stdexcept>

#include "turing-tape.hpp"
#include "turing-transition-function.hpp"

using namespace turing;

turing::turing_transition_function::turing_transition_function(std::string halt_state)
{
    this->halt_state = halt_state;
}

void turing_transition_function::set(turing_transition_function_from frm, turing_transition_function_to to)
{
    if (transitions.find(frm) != transitions.end())
    {
        throw std::invalid_argument("Transition already added");
    }

    transitions[frm] = to;
}

turing_transition_function_to turing_transition_function::get(turing_transition_function_from frm) const
{
    auto transition = transitions.find(frm);
    if (transition == transitions.end())
    {
        return turing_transition_function_to{halt_state, frm.symbol, turing_tape_move::None};
    }

    return transition->second;
}

inline bool turing::operator<(const turing_transition_function_from &lhs, const turing_transition_function_from &rhs)
{
    return (lhs.status < rhs.status) || (lhs.status == rhs.status && lhs.symbol < rhs.symbol);
}